src/tequila/simulators/simulator_ddsim.py

```diff
@@ -0,0 +1,422 @@
+            the gate to be added to the circuit.
+        circuit:
+            the circuit, to which a gate is to be added.
+        args
+        kwargs
+
+        Returns
+        -------
+        None
+        """
+        op = self.op_lookup[gate.name]
+        ddsim_gate = StandardOperation(
+            controls={Control(self.qubit(c))
+                      for c in gate.control},
+            targets=[self.qubit(t) for t in gate.target],
+            op_type=op,
+        )
+        circuit.append(ddsim_gate)
+        self.circuit_sym.append(ddsim_gate)
+
+    def add_measurement(self, circuit, target_qubits, *args, **kwargs):
+        """
+        Add a measurement operation to a circuit.
+        Parameters
+        ----------
+        circuit:
+            a circuit, to which the measurement is to be added.
+        target_qubits: List[int]
+            abstract target qubits
+        args
+        kwargs
+
+        Returns
+        -------
+        circuit with measurements
+
+        """
+        meas = NonUnitaryOperation(targets=target_qubits,
+                                   classics=target_qubits)
+        circuit.append(meas)
+        self.circuit_sym.append(meas)
+        return circuit
+
+    # Overwriting `sample_paulistring` since mqt.ir QuantumComputation object is not pickable:
+    # copy.deepcopy fails.
+    def sample_paulistring(self,
+                           samples: int,
+                           paulistring,
+                           variables,
+                           initial_state: Union[int, QubitWaveFunction] = 0,
+                           *args,
+                           **kwargs) -> numbers.Real:
+        """
+        Sample an individual pauli word (pauli string) and return the average result thereof.
+        Parameters
+        ----------
+        samples: int:
+            how many samples to evaluate.
+        paulistring:
+            the paulistring to be sampled.
+        args
+        kwargs
+
+        Returns
+        -------
+        float:
+            the average result of sampling the chosen paulistring
+        """
+        not_in_u = [
+            q for q in paulistring.qubits if q not in self.abstract_qubits
+        ]
+        reduced_ps = paulistring.trace_out_qubits(qubits=not_in_u)
+        if reduced_ps.coeff == 0.0:
+            return 0.0
+        if len(reduced_ps._data.keys()) == 0:
+            return reduced_ps.coeff
+
+        # make basis change and translate to backend
+        basis_change = QCircuit()
+        qubits = []
+        for idx, p in reduced_ps.items():
+            qubits.append(idx)
+            basis_change += change_basis(target=idx, axis=p)
+
+        # add basis change to the circuit
+        # (deepcopy can't be used with this backend)
+        # can be circumvented by optimizing the measurements
+        # on construction: tq.ExpectationValue(H=H, U=U, optimize_measurements=True)
+        circuit = self.create_circuit(circuit=self.circuit,
```

Comment:
potential problem here, with shallow copy error
(in this case: not variable update is the problem, but accumulation of measurements)